src/WatchDogs.Domain/SuspiciousDealDetector.cs

```diff
@@ -0,0 +1,279 @@
+﻿using Microsoft.Extensions.Options;
+using Serilog;
+using System.Collections.Concurrent;
+using WatchDogs.Contracts;
+using ILogger = Serilog.ILogger;
+
+namespace WatchDogs.Domain;
+
+public class SuspiciousDealDetector : ISuspiciousDealDetector
+{
+    private const decimal VolumeToBalanceTolerance = 0.05M;
+    private static readonly TimeSpan TimeDifferTolerance = TimeSpan.FromSeconds(1);
+    private readonly IDataLoader _dataLoader;
+    private readonly SuspiciousDealDetectorOptions _suspiciousDealDetectorOptions;
+    private readonly ILogger _logger;
+
+    public ConcurrentDictionary<string, ICurrencyBucket> _currencyTradesPairs = new ConcurrentDictionary<string, ICurrencyBucket>();
+    public SuspiciousDealDetector(IDataLoader dataLoader, IOptions<SuspiciousDealDetectorOptions> options, ILogger logger)
+    {
+        _dataLoader = dataLoader;
+        _suspiciousDealDetectorOptions = options.Value;
+        _logger = logger;
+    }
+
+    public async Task<List<Trade>> LoadDealsAsync()
+    {
+        return await _dataLoader.LoadAllTradesAsync();
+    }
+
+    public async Task<List<List<Trade>>> DetectSuspiciousDealsAsync(List<Trade> trades)
+    {
+        List<List<Trade>> dealsAlreadyFilteredByCurrencyPairAndTimeStamp = new();
+
+        await SortTradesByCurrencyPairsAsync(trades);
+
+        var dealsInBuckets = _currencyTradesPairs.Values;
+
+        Log.Information($"Filtering trades into currency buckets. \n" +
```

Comment:
Why `Log` and not `_logger`?

Reply:
I missed that. Mistake made by negligence.

Reply:
Modified